src/crimson/osd/heartbeat.h

```diff
@@ -91,13 +89,28 @@ class Heartbeat : public crimson::net::Dispatcher {
   seastar::future<> send_heartbeats();
   void heartbeat_check();
 
-  struct failure_info_t {
```

Comment:
i don't understand the motivation of this cleanup. IMHO, the existing code is easier to follow.

Reply:
We currently are missing some important implementations around this failure tracking. I was trying to encapsulate the related logic as a self-contained service in order to provide the following interfaces (see the draft implementation in https://github.com/cyx1231st/ceph/blob/wip-seastar-heartbeat-failure-pending/src/crimson/osd/heartbeat.h#L93-L113):
* start to track a peer which is failing but not recognized by monitor: `add_pending(peer...)`;
* stop tracking the peer when the failure report is accepted: `remove_pending(peer)`;
* cancel tracking:
   * `cancel_one()`
   * `cancel_all()`
   * `cancel_if()`

Another reason is to reduce replicated code. Some of them are called from multiple places according to classic OSD, and some are called by the inner `Peer` class.

Reply:
i see. will need to take another look.